### PR TITLE
Cleric Bead Carving

### PR DESCRIPTION
**Pull request by wstampley**
The version of bead carving in Theurgy isn't up to date with the mech lore swap, and doesn't support all the edge cases. Bead carving isn't necessarily a great way to train theurgy itself, but can be useful for field replacing a lost OM orb, favor orbs, etc.

@Hiinky - Any thoughts? 

If you wanted to switch out the bead carving portion of your test-theurgy improvements, you could just call this to make the bead for you.

### CODE REVIEW

File: carve-bead.lic [hunk 1]
```diff
@@ -0,0 +1,152 @@
+=begin
+  Carves a prayer bead from scratch.
```
Comment (rpherbig) [inline carve-bead.lic:2]: Ideally, a lot of this info would live in elanthipedia

**Comment by rpherbig** [comment]
A lot of this logic already exists in https://github.com/rpherbig/dr-scripts/blob/acc89c9a453fcb5afe06e4fa36e31e8b7b8bd4dd/commune.lic#L175 - is there some way to commonize it?

Or is the intent to make `commune.lic` call this script at some point in the future?

File: carve-bead.lic (hunk 1 above)
Comment (wstampley) [reply carve-bead.lic:2]: I can add a section for this script if we merge it

**Comment by Hiinky** [comment]
Sorry ahead of time for the novel:

A common theurgy might be the easiest way to go about it.  Honestly a lot of clerics train theurgy a little differently and having common methods could be very useful.  In that same train of thought, I haven't done the research yet but I want to add support for All God that definitely does not fit with the current script and moving to a common would avoid even more potential duplication when I go to work on that project.  Without looking, I would guess the information in commune.lic is also out of date and would need the same update as theurgy's carving section so having commune use this script (and/or a common) would make sense.

I have all kinds of ideas for a common cleric-y things that might be awesome, if you want to DM me and make a bigger project of it.  For example it might also be appreciated by clerics everywhere to include a yaml field/support for the truffenyi commune with options to use it for getting/returning favor orbs for their chosen deity.  

For this PR specifically, I took a poll in discord and the vast majority of responses wanted support for bead carving and for practical use, it should be available to call independent from normal theurgy training.  Whether it's by an arg in theurgy or a separate script, I'm not picky about which way it goes.

File: commune.lic [hunk 2]
```diff
@@ -1,310 +1,27 @@
...
-    case bput("commune truffenyi #{target}", 'The power of Truffenyi has answered your prayer', 'You must be holding your favor offering or orb in your right hand', @commune_errors, @devotion_levels)
-    when 'The power of Truffenyi has answerd your prayer'
-      return true
-    when 'You must be holding your favor offering or orb in your right hand'
-      missing_item('a proper offering in your right hand')
-    else
-      return false
+  def initialize
+    if !DRStats.cleric?
+      DRC.message('Silly rabbit, communes are for clerics.')
```
Comment (KatoakDR) [inline commune.lic:13]: 🤣 🐰

File: carve-bead.lic [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        { name: 'aspect', regex: /^[A-z0-9\s\-\'\.]+$/i, optional: true, description: 'Aspect to use, regardless of settings' }
+      ]
+    ]
+    args = parse_args(arg_definitions)
+
+    @settings = get_settings
+    @debug = args.debug || @settings['carve-bead']['debug']
+    @theurgy_supply_container = @settings.theurgy_supply_container
+    @immortal_aspect = args.aspect || @settings.immortal_aspect
+    @bag = @settings.crafting_container
```
[the remark this answers is not in the document]
Comment (wstampley) [reply carve-bead.lic:45]: Makes it easier to follow. Pretty sure I stole a bunch of that, names and all, from one of the actual crafting scripts.

File: carve-bead.lic [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    @bag_items = @settings.crafting_items_in_container
+    @belt = @settings.engineering_belt
+    @water_holder = @settings.water_holder
+    @use_primer = args.primer || @settings['carve-bead']['primer'] || false
```
[the remark this answers is not in the document]
Comment (wstampley) [reply carve-bead.lic:49]: While some part of my brain knows that 'nil' is falsey and weakly typed languages exist, I tend to be extra explicit. I do this with the (weird to me) auto-returns ruby functions have as well.

File: carve-bead.lic [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+  def main
+    DRCI.stow_hands
+    read_primer if @use_primer
+    make_block
+    carve_bead
+  end
+
+  def read_primer
+    DRCI.get_item_safe("#{@immortal_aspect} primer", @theurgy_supply_container)
```
[the remark this answers is not in the document]
Comment (wstampley) [reply carve-bead.lic:76]: Are the _safe methods designed to be internals to the module? I wonder if there's a way to enforce that to make the interface cleaner.

File: carve-bead.lic [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    DRCI.put_away_item?("#{@immortal_aspect} primer", @theurgy_supply_container)
+  end
+
+  def open_primer
+    DRC.bput("open my primer", "You open the cover of", "The primer is already open")
+  end
+
+  def close_primer
+    DRC.bput("close my primer", ".*")
+  end
```
[the remark this answers is not in the document]
Comment (wstampley) [reply carve-bead.lic:107]: I think that's just confusing though, cause it's not a container, even though it matches the final command, it's a book.

File: carve-bead.lic [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+      else
+        keep_reading = false
+      end
+    end
+    DRCI.put_away_item?("#{@immortal_aspect} primer", @theurgy_supply_container)
+  end
+
+  def open_primer
+    DRC.bput("open my primer", "You open the cover of", "The primer is already open")
+  end
```
[the remark this answers is not in the document]
Comment (wstampley) [reply carve-bead.lic:103]: I think that's just confusing though, cause it's not a container, even though it matches the final command, it's a book.

File: carve-bead.lic [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    DRC.bput("close my primer", ".*")
+  end
+
+  def reset_primer
+    close_primer
+    open_primer
+  end
+
+  def make_block
+    return false unless DRCTH.has_holy_water?(@theurgy_supply_container, @water_holder)
```
[the remark this answers is not in the document]
Comment (wstampley) [reply carve-bead.lic:115]: holy water is usually more complicated than that, because it's a container in a container. The common-theurgy method is trying to handle all that (and might rely on the DRCI method).

File: carve-bead.lic [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+    DRCT.walk_to @data['limb_foraging']['id']
+    DRCA.prepare?('bless', 1)
+    DRC.forage? 'limb'
+    DRCI.get_item_safe(@water_holder, @theurgy_supply_container)
+    DRC.bput("sprinkle #{@water_holder} on my limb", 'You sprinkle')
```
[the remark this answers is not in the document]
Comment (wstampley) [reply carve-bead.lic:121]: You shouldn't be able to get here with empty water holder, because of the 'has holy water' check at the top.

File: carve-bead.lic [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    DRCI.put_away_item?(@water_holder, @theurgy_supply_container)
+    pause 1
+    DRCA.cast?('cast my limb')
```
[the remark this answers is not in the document]
Comment (wstampley) [reply carve-bead.lic:124]: A 1 mana bless is snap-castable for almost any character, common forage roundtimes for successes are what, 4 seconds?

 I will likely be refactoring this to the DRCTH method that was extracted from 'commune', and might make that configurable.

File: common-items.lic [hunk 11]
```diff
@@ -294,6 +294,18 @@ module DRCI
     end
   end
 
+  def have_item_by_look?(item, container)
+    item = "my #{item}" if item && !(item =~ /^my /i)
+    container = "my #{container}" if container && !(container =~ /^((in|on|under|behind|from) )?my /i)
+
+    case DRC.bput("look at #{item} in #{container}", 'You see nothing unusual', 'I could not find', 'What were you referring to')
```
[the remark this answers is not in the document]
Comment (wstampley) [reply common-items.lic:301]: Are you suggesting the final command be `look at item from my container`? That doesn't make sense to me.

File: common-theurgy.lic [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+=end
+
+custom_require.call(%w[common common-items])
+
+module DRCTH
+  module_function
+
+  CLERIC_ITEMS = [
+    'holy water', 'holy oil', 'wine', 'incense', 'flint', 'chamomile', 'sage', 'jalbreth balm'
+  ] unless defined?(CLERIC_ITEMS)
```
[the remark this answers is not in the document]
Comment (wstampley) [reply common-theurgy.lic:12]: The unit tests (and probably other situations) load and reload this module over and over, which causes warnings about constant redefinition.

I'm not sure what that says about my understanding of how modules get loaded in Ruby/Lich, but for now this fixes it and gives me the behavior I want.

File: common-theurgy.lic [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+    end
+
+    return unless shop_data['needs_bless'] && @known_spells.include?('Bless')
+    quick_bless_item(item[:name])
+  end
+
+  def quick_bless_item(item_name)
+    # use dummy settings object since this isn't complex enough for camb, etc.
+    DRCA.cast_spell({
+        'abbrev' => 'bless', 'mana' => 1, 'prep_time' => 2, 'cast' => "cast my #{item_name}"
```
[the remark this answers is not in the document]
Comment (wstampley) [reply common-theurgy.lic:106]: Not sure. This was extracted from existing code, and I didn't see any open issues for it, so I left the timer as is.

File: common-theurgy.lic [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+  def empty_cleric_hands(theurgy_supply_container)
+    empty_cleric_right_hand(theurgy_supply_container)
+    empty_cleric_left_hand(theurgy_supply_container)
+  end
+
+  def empty_cleric_right_hand(theurgy_supply_container)
+    return if GameObj.left_hand.noun.nil?
+
+    container = CLERIC_ITEMS.include?(GameObj.right_hand.noun) || CLERIC_ITEMS.include?(GameObj.right_hand.name) ? theurgy_supply_container : nil
+    DRCI.put_away_item_safe(GameObj.right_hand.noun, container)
```
[the remark this answers is not in the document]
Comment (wstampley) [reply common-theurgy.lic:119]: Extracted from existing code. I'm happy to try this version of it and see.

File: common-theurgy.lic [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+  end
+
+  def empty_cleric_left_hand(theurgy_supply_container)
+    return if GameObj.left_hand.noun.nil?
+
+    container = CLERIC_ITEMS.include?(GameObj.left_hand.noun) || CLERIC_ITEMS.include?(GameObj.left_hand.name) ? theurgy_supply_container : nil
+    DRCI.put_away_item_safe(GameObj.left_hand.noun, container)
```
[the remark this answers is not in the document]
Comment (wstampley) [reply common-theurgy.lic:126]: My code was using _safe (cause I thought I needed it). Extracted code from other scripts was probably using the other version. I've standardized on the (un-safe) method.

File: common-theurgy.lic [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+    end
+    if !DRCI.get_item_safe('incense', theurgy_supply_container)
+      DRC.message("Can't find incense to light")
+      empty_cleric_hands(theurgy_supply_container)
+      return false
+    end
+
+    while DRC.bput('light my incense with my flint', 'nothing happens', 'bursts into flames', 'much too dark in here to do that') == 'nothing happens'
+      waitrt?
+    end
```
[the remark this answers is not in the document]
Comment (wstampley) [reply common-theurgy.lic:161]: I've never had it happen, but it's probably worth defending against.

File: carve-bead.lic (hunk 5 above)
[the remark this answers is not in the document]
Comment (KatoakDR) [reply carve-bead.lic]: I don't think it's so much that they're designed to be internals as much as the (my?) "preferred" method to use is the less verbose `get_item?` alias. In my mind, I can see a developer maybe using the `_unsafe` variant if they had a use case where they really needed the script to interact with something that's literally not on their person/inventory. The only use case I can think of would be like a vault or home furniture?

File: carve-bead.lic (hunk 7 above)
[the remark this answers is not in the document]
Comment (KatoakDR) [reply carve-bead.lic]: ok

File: carve-bead.lic (hunk 6 above)
[the remark this answers is not in the document]
Comment (KatoakDR) [reply carve-bead.lic]: ok

File: common-items.lic (hunk 11 above)
[the remark this answers is not in the document]
Comment (KatoakDR) [reply common-items.lic:301]: Heh, the grammar part wasn't the objective of my suggestion. Sorry, I should have clarified. The change is to not hardcode the prepositions (in, from, under, ..) in the command so that the preposition optionally can be passed in as part of the argument.

Here's some examples of what I'm talking about:
```
DRCI.have_item_by_look?('rope', 'pack')         # look at my rope in my pack
DRCI.have_item_by_look?('rope', 'in drawer')    # look at my rope in drawer
DRCI.have_item_by_look?('rope', 'behind table') # look at my rope behind table
DRCI.have_item_by_look?('rope', 'on my shelf')  # look at my rope on my shelf
```

Does that make more sense?

As an aside, although in plain english the phrase "look at item from my container" doesn't read well, the game engine parses it fine.
```
> look my pouch from my pack
The soft gem pouch is made for easy storage of gems, and is used by gemshops everywhere to handle bulk transactions.
```

Thanks

[the remark this answers is not in the document]
Comment (KatoakDR) [reply common-items.lic:301]: Another thought, to expand this method's utility to checking if an item is on your person (worn or stowed), then can modify the container argument in the signature to be `container = nil`.

Also, since this method uses the `look` verb, if an item has a descriptive look then it needs to handle more match strings. Since the permutations are boundless for what items LOOK like, but are fairly limited for when an item can't be found, I suggest this change:

```ruby
  def have_item_by_look?(item, container = nil)
    item = "my #{item}" if item && !(item =~ /^my /i)
    container = "in my #{container}" if container && !(container =~ /^((in|on|under|behind|from) )?my /i)

    case DRC.bput("look at #{item} #{container}", 'I could not find', 'What were you referring to', '.*')
    when 'I could not find', 'What were you referring to'
      false
    else
      true
    end
  end
```

Here's some quick tests:
```
>,e echo DRCI.have_item_by_look?('hairbrush')

[exec1]>look at my hairbrush 
I could not find what you were referring to.
> 
[exec1: false]

>,e echo DRCI.have_item_by_look?('soft pouch')

[exec1]>look at my soft pouch 
The soft gem pouch has been tied off.
The soft gem pouch is made for easy storage of gems, and is used by gemshops everywhere to handle bulk transactions.
> 
[exec1: true]

>,e echo DRCI.have_item_by_look?('pack')
[exec1]>look at my pack 
The steelsilk is woven through the fabric in such a way to form the image of a broad tower shield.
> 
[exec1: true]
```

[the remark this answers is not in the document]
Comment (KatoakDR) [reply common-items.lic:301]: Another thought is use the existing `tap(item, container = nil)` method. Would that work for your use case and thus not need this new method?

File: carve-bead.lic (hunk 9 above)
[the remark this answers is not in the document]
Comment (KatoakDR) [reply carve-bead.lic:119]: True. Doh!

File: test/test_common_theurgy.rb [hunk 17]
```diff
@@ -0,0 +1,130 @@
+require 'minitest/autorun'
+load 'test/test_harness.rb'
+
+include Harness
+
+class TestPerceiveHealth < Minitest::Test
```
Comment (KatoakDR) [inline test/test_common_theurgy.rb:6]:
```suggestion
class TestCommonTheurgy < Minitest::Test
```

File: common-items.lic (hunk 11 above)
[the remark this answers is not in the document]
Comment (wstampley) [reply common-items.lic:301]: The thing about tap is that it's very RP breaking, in certain spots where you need to verify a lot of things at once. Like, it's a dead giveaway to me that someone else is using lich when they stand up, tap 5 things, and run out the room.

I like the silent 'look' verify where it works.

[the remark this answers is not in the document]
Comment (wstampley) [reply common-items.lic:301]: That said, I'm all for switching to a negative match test, and fixing the adverbs like you suggested.

[the remark this answers is not in the document]
Comment (wstampley) [reply common-items.lic:301]: One downside to .* matching is that it will match any game response, regardless of if it was actually the response to look, which could mean false positives on having things in noisy rooms. Thoughts?

[the remark this answers is not in the document]
Comment (KatoakDR) [reply common-items.lic:301]: Yeah, that’s definitely a downside and entirely possible that an unrelated game response is received before a “What we’re you referring to?”

If the noun of the item you’re trying to look at will be in the LOOK then the .* maybe can be changed to something akin to `/\b#{item.strip.split(' ').last}\b/`

File: test/test_common_theurgy.rb (hunk 17 above)
Comment (KatoakDR) [reply test/test_common_theurgy.rb:6]: Curious, this suggestion is marked resolved but the test file still shows as `TestPerceiveHealth`. Is the commit to perform the rename done locally and yet to be pushed? Thanks

**Comment by rpherbig** [comment]
What's left before this is ready to be merged?

**Comment by wstampley** [comment]
> What's left before this is ready to be merged?

@rpherbig - The vast majority of the changes in this PR are good to go. I would want to put some final cleanup and testing into what I have here if we were to do that.

However, I am still (slowly) going through and moving things to use the common methods presented here.

If we want to get this one closed, and come back to update scripts like Theurgy to use bead-carving and common-theurgy methods, I'm okay with that.

**Comment by rpherbig** [comment]
> However, I am still (slowly) going through and moving things to use the common methods presented here.
> If we want to get this one closed, and come back to update scripts like Theurgy to use bead-carving and common-theurgy methods, I'm okay with that.

The barrier for adding a new script is really low. It's nearly impossible for it to break an existing user's configuration, which is really the biggest risk to merging changes.

Commonizing code and cleanup is great, but shouldn't hold up adding a new script. I encourage you to get this to a state where we can get the new script into users hands and it's fine to then come back and make a new common-theurgy.

**Comment by wstampley** [comment]
Everything you can reach right now is working:
- all the communes in commune.lic
- all the bead carving stuff in carve-bead.lic
- any common-theurgy methods that either of them call

I didn't get to:
- fully relying on the common methods in carve-bead
- switching stuff in therugy.lic over to the common-methods
- full unit test coverage for the stuff that was moved

**Comment by wstampley** [comment]
@rpherbig @KatoakDR @Hiinky - I flipped this to ready to review again, there is a comment in the thread about the state of everything.

**Comment by rpherbig** [comment]
These changes look fine to me, but it's pretty hard to review.

@KatoakDR @Hiinky let me know what you're thinking...

File: common-theurgy.lic [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+    item_shop_data = town_theurgy_data["#{item_name}_shop"]
+    return false if item_shop_data.nil?
+
+    DRCT.walk_to(item_shop_data['id'])
+    if stackable
+      num_to_buy.times do
+        buy_single_supply(item_name, item_shop_data)
+        if DRCI.get_item?(item_name, theurgy_supply_container)
+          DRC.bput("combine #{item_name} with #{item_name}",
+               'You combine', 'You can\'t combine', 'You must be holding')
```
Comment (KatoakDR) [inline common-theurgy.lic:81]: Not sure if the newline was on purpose?
```suggestion
          DRC.bput("combine #{item_name} with #{item_name}", 'You combine', 'You can\'t combine', 'You must be holding')
```

**Comment by KatoakDR** [comment]
> These changes look fine to me, but it's pretty hard to review.
> 
> @KatoakDR @Hiinky let me know what you're thinking...

It's definitely a big PR. Any pure code things look fine to me, too. I don't play a Cleric or have ever used these scripts so unfortunately I'm not a good candidate to help sanity test the scripts.

@wstampley Have any other Cleric players been recruited to help test?

Thanks

File: common-theurgy.lic (hunk 18 above)
Comment (wstampley) [reply common-theurgy.lic:81]: Eh, it just splits up the matches from the command for readability on the long line, but it doesn't matter.

**Comment by wstampley** [comment]
> @wstampley Have any other Cleric players been recruited to help test?
> 
> Thanks

A couple clerics tested carve-bead along the way. I'll put out a call in Discord to see if anyone wants to test.

It's a pain to test for most people, though, cause it touches many files and several require reloading common, etc.

**Comment by wstampley** [comment]
Testers were requested on the 10th:
https://discord.com/channels/745675889622384681/745676326190579843/852745616542793788

No one has sent me any feedback, etc. Happy to give it more time for people to test, or whatever you all think is best.
